cpp/include/cudf/ast/detail/expression_evaluator.cuh

```diff
@@ -595,31 +596,42 @@ struct expression_evaluator {
      * @param output Output data reference.
      */
     template <ast_operator op,
-              typename OutputType,
+              typename ResultSubclass,
+              typename T,
+              bool result_has_nulls,
               std::enable_if_t<
                 detail::is_valid_unary_op<detail::operator_functor<op, has_nulls>,
                                           possibly_null_value_t<Input, has_nulls>>>* = nullptr>
-    __device__ void operator()(OutputType& output_object,
-                               cudf::size_type const output_row_index,
-                               possibly_null_value_t<Input, has_nulls> const input,
-                               detail::device_data_reference const output) const
+    CUDA_DEVICE_CALLABLE void operator()(
+      expression_result<ResultSubclass, T, result_has_nulls>& output_object,
+      cudf::size_type const output_row_index,
+      possibly_null_value_t<Input, has_nulls> const& input,
+      detail::device_data_reference const& output,
+      IntermediateDataType<has_nulls>* thread_intermediate_storage) const
     {
       // The output data type is the same whether or not nulls are present, so
       // pull from the non-nullable operator.
       using Out = cuda::std::invoke_result_t<detail::operator_functor<op, false>, Input>;
-      this->template resolve_output<Out>(
-        output_object, output, output_row_index, detail::operator_functor<op, has_nulls>{}(input));
+      this->template resolve_output<Out>(output_object,
+                                         output,
+                                         output_row_index,
+                                         thread_intermediate_storage,
+                                         detail::operator_functor<op, has_nulls>{}(input));
     }
 
     template <ast_operator op,
-              typename OutputType,
+              typename ResultSubclass,
+              typename T,
+              bool result_has_nulls,
               std::enable_if_t<
                 !detail::is_valid_unary_op<detail::operator_functor<op, has_nulls>,
                                            possibly_null_value_t<Input, has_nulls>>>* = nullptr>
-    __device__ void operator()(OutputType& output_object,
-                               cudf::size_type const output_row_index,
-                               possibly_null_value_t<Input, has_nulls> const input,
-                               detail::device_data_reference const output) const
+    CUDA_DEVICE_CALLABLE void operator()(
+      expression_result<ResultSubclass, T, result_has_nulls>& output_object,
+      cudf::size_type const output_row_index,
+      possibly_null_value_t<Input, has_nulls> const& input,
+      detail::device_data_reference const& output,
+      IntermediateDataType<has_nulls>* thread_intermediate_storage) const
     {
       cudf_assert(false && "Invalid unary dispatch operator for the provided input.");
```

Comment:
The binary version of this was removed in 8ee0d5451eb14213f87b7e1c18129a832a18e901. Should the unary one be removed as well?

Reply:
What was removed in that commit was an `operator()` of the `expression_evaluator`. What you're pointing out here is an `operator()` of `expression_evaluator::unary_expression_output_handler`. This function definition _should_ exist (as well as the one for `expression_evaluator::binary_expression_output_handler`) because this is the place in the call stack where we are constructing the `operator_functor` and therefore can choose the appropriate overload based on the validity of the operation. The operator that was removed could never be called, because at that point the input types would not yet be resolved so there was no way to select the appropriate overload based on whether the `operator_functor` could be invoked with those types. I think it was probably just some cruft from your early implementations that was never removed.

Reply:
I misread this. The error messages and `operator()` were identical. Sorry!